source/extensions/filters/listener/original_src/original_src_socket_option.cc

```diff
@@ -0,0 +1,70 @@
+#include "extensions/filters/listener/original_src/original_src_socket_option.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace OriginalSrc {
+
+constexpr uint8_t OriginalSrcSocketOption::IPV4_KEY;
+constexpr uint8_t OriginalSrcSocketOption::IPV6_KEY;
+
+OriginalSrcSocketOption::OriginalSrcSocketOption(
+    Network::Address::InstanceConstSharedPtr src_address)
+    : src_address_(std::move(src_address)) {
+  // Source transparency only works on IP connections.
+  ASSERT(src_address_->type() == Network::Address::Type::Ip);
+}
+
+bool OriginalSrcSocketOption::setOption(
+    Network::Socket& socket, envoy::api::v2::core::SocketOption::SocketState state) const {
+
+  if (state == envoy::api::v2::core::SocketOption::STATE_PREBIND) {
+    socket.setLocalAddress(src_address_, false);
+  }
+
+  bool result = true;
+  std::for_each(options_to_apply_.begin(), options_to_apply_.end(),
+                [&socket, state](const Network::Socket::OptionConstSharedPtr& option) {
+                  option->setOption(socket, state);
+                });
+  return result;
+}
+
+/**
+ * Inserts an address, already in network order, to a byte array.
+ */
+template <typename T> void addressIntoVector(std::vector<uint8_t>& vec, const T& address) {
+  const uint8_t* byte_array = reinterpret_cast<const uint8_t*>(&address);
+  vec.insert(vec.end(), byte_array, byte_array + sizeof(T));
+}
+
+void OriginalSrcSocketOption::hashKey(std::vector<uint8_t>& key) const {
+  /* We do two things here to ensure that there isn't any ambiguity when combining the hash key with
+   * variable length options:
+   * 1. Ensure the the key is fixed length by padding ipv4 addresses up to 16 bytes with 0s.
+   * 2. Enusre that Ipv6 addresses cannot collide with padded v4 addresses by placing a unique value
+   *    before the address.
+   */
+  if (src_address_->ip()->version() == Network::Address::IpVersion::v4) {
+    key.push_back(IPV4_KEY);
```

Comment:
I recall the connection pools are per host, so they should already be separated by the destination address. If so, the insertion of the address family into the key should not be necessary.

Reply:
The source address is what I'm concerned about here. I guess the question is whether or not we'll have an ipv6 and ipv4 request going to the same upstream host? If it's dual stack, it'd work. Does envoy support dual-stack?

I'll give you an example of what I'm concerned with.

Suppose we have two connections 1.2.3.4 and 0::0:0:12:34 to the same upstream host. Now imagine that we have another variable length hash key inserted before the original_src key. In the IPv4 case, it has a 13 byte value 
`0x01 0000 0000 0000 0000 0000 0000 # 1 byte of value 1, followed by 12 bytes of 0`

Now imagine that in the IPv6 case, it has value `0x01`, because it's also variable length. If you line them up, you'll get:
```
0x01 0000 0000 0000 0000 0000 0000 0102 0304 # ipv4
0x01 0000 0000 0000 0000 0000 0000 0102 0304 # ipv6
```
That is, the hash keys are the same, for very different IP addresses. Both will go to the same connection pool, which will lead to problems.

If the situation where Ipv4 and Ipv6 go to the same upstream host is impossible, then we definitely don't have a problem. Otherwise, yeah, I'm being a little paranoid, but is it so bad? :)

Reply:
> I guess the question is whether or not we'll have an ipv6 and ipv4 request going to the same upstream host?

While Envoy support dual-stack and it is possible but I think they will end up in different listening socket so it should be fine. You might use [IPv6 mapped IPv4 address](https://tools.ietf.org/html/rfc4291#section-2.5.5.2) for this purpose.

Reply:
@lizan I thought about doing that, but I was concerned about conflicts between upstreams using actually mapped addresses (e.g. an upstream supporting only v6), and upstreams using the v4 one.

Either way, if we think it's impossible, I can just remove it. It's cleaner that way.

Reply:
> I was concerned about conflicts between upstreams using actually mapped addresses (e.g. an upstream supporting only v6), and upstreams using the v4 one.

In that case I think kernel will convert the v4 address to IPv6 mapped IPv4 address so it should be fine with that, am I missed anything?

Reply:
Hmmm.. So, the case to be concerned with would be:

Two downstream connections, one from ipv6 mapped a.b.c.d, and one from plain ipv4 a.b.c.d. They are both sent to an upstream ipv6 host. The ipv6 mapped a.b.c.d uses the original address for its upstream connection. The ipv4 connection uses the ipv6 mapped one since the linux kernel maps it for us. In that case, we *should* use the same connection pool. However, we won't with the current code, because I have a different hash for ipv4 and ipv6, even if they are the same underlying ipv4 address.

I can change this to use IPv6 mapped Ipv4, then.

Reply:
I'm fine either way as long as we don't use same hash for downstream connections shouldn't be same. Just leave some notes as comment?

Reply:
The hash can't be the same. I tested this out, and it looks like you cannot bind a v4 address onto a v6 socket. Further, I was unable to bind a v4-mapped v6 address onto a v6 socket.  So, I think this means the case I'm concerned about can't happen. Basically, v4 can only go to a v4 host, and v6 can only go to v6 host.